File: functions/forecastingSARMA/test/getAutoArimaModelCoefficientsAndForecast.R
# Read in historic data (and other info)
historicData <- read.table('historicData.csv', header=FALSE, sep=',')
horizon <- read.table('intervalsToForecast.csv', header=FALSE, sep=',')
seasonality <- read.table('seasonality.csv', header=FALSE, sep=',')
order <- read.table('order.csv', header=FALSE, sep=',')
seasonalOrder <- read.table('seasonalOrder.csv', header=FALSE, sep=',')

# Load the forecast package:
library(forecast)

# Cast historicData as time-series
historicDataTimeSeries <- ts(historicData$V1, frequency=seasonality$V1);

# Carry out automated forecast
arimaFit <- Arima(historicDataTimeSeries, order=order$V1, seasonal=seasonalOrder$V1, 
                       include.mean=FALSE);

arimaForecast <- forecast(arimaFit, h=horizon$V1);

# Save the mean forecast as CSV file
write.table(arimaForecast$mean, file='meanForecast.csv', sep=',',
                col.names=FALSE, row.names=FALSE, qmethod='double')

# Save the model coefficients as CSV file
write.table(as.numeric(arimaFit$coef), file='coefficients.csv', sep=',',
            col.names=FALSE, row.names=FALSE, qmethod='double')

File: functions/forecastingSARMA/test/getAutoArimaModelCoefficientsAndForecast.m
function [coefficients, meanForecast] = ...
    getAutoArimaModelCoefficientsAndForecast(cfg, dataValues, ...
    order, seasonalOrder)

%% Create temprary directory for storing temporary files (and move into it)
originalDir = pwd;
tmpName = tempname;
mkdir(tmpName);
[locationOfRfile, ~, ~] = fileparts(which(...
    'getAutoArimaModelCoefficientsAndForecast.m'));
copyfile([locationOfRfile...
     filesep 'getAutoArimaModelCoefficientsAndForecast.R'],tmpName);
cd(tmpName);

%% Write historic (& other info) as columns
csvwrite('historicData.csv', dataValues(:));
csvwrite('seasonality.csv', cfg.fc.season);
csvwrite('intervalsToForecast.csv', cfg.sim.horizon);
csvwrite('order.csv', order(:));
csvwrite('seasonalOrder.csv', seasonalOrder(:));

%% Call the R forecasting script:
system('R CMD BATCH getAutoArimaModelCoefficientsAndForecast.R debug.txt');

%% Read in the mean forecast & coefficients (saved by the R script):
meanForecast = csvread('meanForecast.csv');
coefficients = csvread('coefficients.csv');

%% Return to rigina directlroy and destroy temporary directory
cd(originalDir);
rmdir(tmpName, 's');

end
